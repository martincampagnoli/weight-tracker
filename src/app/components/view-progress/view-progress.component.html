<p-chart
  type="line"
  [data]="weightChartData"
  [options]="weightChartOptions"
  styleClass="w-full"
>
</p-chart>
