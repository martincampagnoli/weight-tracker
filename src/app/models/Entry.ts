export interface Entry {
  id: number;
  weight: number;
  date: string;
  description?: string;
}
