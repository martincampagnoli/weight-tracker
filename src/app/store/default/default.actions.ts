import { createAction, props } from '@ngrx/store';
import { Entry } from 'src/app/models/Entry';

/**
 * Action representing a failure during data retrieval.
 */
export const getFailure = createAction(
  '[Default] GetFailure  Error',
  props<{ payload: { message: string } }>()
);

/**
 * Action to initiate retrieval of data.
 */
export const getData = createAction('[Entry] Get Entries');

/**
 * Action indicating successful retrieval of data.
 */
export const getDataSuccess = createAction(
  '[Entry] Get Entries Success',
  props<{ payload: Entry[] }>()
);

/**
 * Action to add a new entry.
 */
export const addEntrySuccess = createAction(
  '[Default] Add Entry Success',
  props<{ payload: { entry: Entry } }>()
);

/**
 * Action to delete a new entry.
 */
export const deleteEntrySuccess = createAction(
  '[Default] Delete Entry Success',
  props<{ payload: { id: number } }>()
);

/**
 * Action to reset the application state.
 */
export const resetAppState = createAction('[Default] Reset app state');
