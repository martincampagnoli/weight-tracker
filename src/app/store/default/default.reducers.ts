import {
  Action,
  createFeatureSelector,
  createReducer,
  createSelector,
  on,
} from '@ngrx/store';
import {
  addEntrySuccess,
  deleteEntrySuccess,
  getDataSuccess,
  resetAppState,
} from './default.actions';
import { Entry } from 'src/app/models/Entry';

/**
 * The structure of the application state.
 */
export interface AppState {
  loading: boolean;
  data: Entry[];
}

/**
 * The default initial state for the application.
 */
const initialState: AppState = {
  loading: false,
  data: [],
};

/**
 * Reducer function that handles state changes based on dispatched actions.
 *
 * @param state The current state.
 * @param action The dispatched action.
 * @returns The new state after applying the action.
 */
const reducer = createReducer(
  initialState,
  on(resetAppState, (initialState) => ({
    ...initialState,
  })),
  on(getDataSuccess, (state, { payload }) => {
    return {
      ...state,
      loading: false,
      data: payload,
    };
  }),
  on(deleteEntrySuccess, (state, { payload }) => {
    return {
      ...state,
      loading: false,
      data: state.data.filter((entry) => entry.id !== payload.id),
    };
  }),
  on(addEntrySuccess, (state, { payload }) => {
    return {
      ...state,
      loading: false,
      data: [...state.data, payload.entry],
    };
  })
);

/**
 * The reducer function that handles various app-related actions to modify the state.
 *
 * @param state The current state.
 * @param action The dispatched action.
 * @returns The new state after applying the action.
 */
export function appReducer(state: AppState | undefined, action: Action) {
  return reducer(state, action);
}

/**
 * Selects the 'app' feature state.
 */
export const selectAppState = createFeatureSelector<AppState>('app');

/**
 * Selectors for various parts of the application state.
 */
export const getAppState = createSelector(
  selectAppState,
  (state: AppState) => state
);

export const getDataState = createSelector(
  selectAppState,
  (state: AppState) => state.data
);
