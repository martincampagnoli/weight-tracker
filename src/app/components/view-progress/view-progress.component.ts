import {
  Component,
  inject,
  OnInit,
  DestroyRef,
  ChangeDetectionStrategy,
} from '@angular/core';
import { CommonModule } from '@angular/common';
import { ChartModule } from 'primeng/chart';
import { STEPPER_GLOBAL_OPTIONS } from '@angular/cdk/stepper';
import { Store } from '@ngrx/store';
import * as reducers from 'src/app/store/default';
import { Entry } from 'src/app/models/Entry';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { DataService } from 'src/app/services/data.service';

interface ChartDataset {
  label: string;
  data: number[];
  fill: boolean;
  borderColor: string;
  tension: number;
  pointRadius: number;
}

interface WeightChartData {
  labels: string[];
  datasets: ChartDataset[];
}

interface ChartOptions {
  responsive: boolean;
  plugins: {
    legend: { display: boolean };
    tooltip: { enabled: boolean };
  };
  scales: {
    x: { title: { display: boolean; text: string } };
    y: { title: { display: boolean; text: string } };
  };
}

@Component({
  selector: 'app-view-progress',
  templateUrl: './view-progress.component.html',
  styleUrls: ['./view-progress.component.scss'],
  standalone: true,
  changeDetection: ChangeDetectionStrategy.OnPush,
  providers: [
    {
      provide: STEPPER_GLOBAL_OPTIONS,
      useValue: { showError: true },
    },
  ],
  imports: [CommonModule, ChartModule],
})
export class ViewProgressComponent implements OnInit {
  private readonly store = inject(Store<reducers.AppState>);
  private destroyRef = inject(DestroyRef);
  private dataService = inject(DataService);

  weightChartData: WeightChartData | null = null;
  weightChartOptions: ChartOptions;

  private readonly CHART_CONFIG = {
    BORDER_COLOR: '#42A5F5',
    TENSION: 0.3,
    POINT_RADIUS: 5,
    WEIGHT_LABEL: 'Weight (kg)',
    DATE_LABEL: 'Date',
  } as const;

  constructor() {
    this.weightChartOptions = this.createChartOptions();
  }

  ngOnInit(): void {
    this.loadWeightData();
  }

  private loadWeightData(): void {
    this.store
      .select(reducers.getDataState)
      .pipe(takeUntilDestroyed(this.destroyRef))
      .subscribe((data) => {
        if (data.length > 0) {
          const sortedData = this.dataService.sortEntriesByDate(data);
          this.weightChartData = this.createChartData(sortedData);
        }
      });
  }

  private createChartData(entries: Entry[]): WeightChartData {
    return {
      labels: entries.map((entry) => entry.date),
      datasets: [
        {
          label: this.CHART_CONFIG.WEIGHT_LABEL,
          data: entries.map((entry) => entry.weight),
          fill: false,
          borderColor: this.CHART_CONFIG.BORDER_COLOR,
          tension: this.CHART_CONFIG.TENSION,
          pointRadius: this.CHART_CONFIG.POINT_RADIUS,
        },
      ],
    };
  }

  private createChartOptions(): ChartOptions {
    return {
      responsive: true,
      plugins: {
        legend: { display: true },
        tooltip: { enabled: true },
      },
      scales: {
        x: {
          title: {
            display: true,
            text: this.CHART_CONFIG.DATE_LABEL,
          },
        },
        y: {
          title: {
            display: true,
            text: this.CHART_CONFIG.WEIGHT_LABEL,
          },
        },
      },
    };
  }
}
