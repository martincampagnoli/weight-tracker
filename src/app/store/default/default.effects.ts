import { Injectable } from '@angular/core';
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { of, catchError, switchMap } from 'rxjs';
import { DataService } from 'src/app/services/data.service';
import { getFailure, getData } from './default.actions';
import { getDataSuccess } from './default.actions';
import { Entry } from 'src/app/models/Entry';

/**
 * Effects for managing application actions related to data.
 */
@Injectable()
export class AppEffects {
  /**
   * Constructs the AppEffects.
   *
   * @param actions$ Observable stream of dispatched actions.
   * @param moviesService The service for fetching data.
   */
  constructor(private actions$: Actions, private dataService: DataService) {}

  /**
   * Effect to fetch data.
   */
  getData = createEffect(() =>
    this.actions$.pipe(
      ofType(getData),
      switchMap(() =>
        this.dataService.getData().pipe(
          switchMap((data: Entry[]) => [getDataSuccess({ payload: data })]),
          catchError((error) => of(getFailure({ payload: error })))
        )
      )
    )
  );
}
