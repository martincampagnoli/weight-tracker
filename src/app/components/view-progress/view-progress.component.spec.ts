import { ComponentFixture, TestBed } from '@angular/core/testing';
import { Store } from '@ngrx/store';
import { MockStore, provideMockStore } from '@ngrx/store/testing';
import { MatDialogRef } from '@angular/material/dialog';
import { signal } from '@angular/core';
import { of } from 'rxjs';

import { ViewProgressComponent } from './view-progress.component';
import { Entry } from 'src/app/models/Entry';
import * as reducers from 'src/app/store/default';

describe('ViewProgressComponent', () => {
  let component: ViewProgressComponent;
  let fixture: ComponentFixture<ViewProgressComponent>;
  let store: MockStore;
  let mockDialogRef: jest.Mocked<MatDialogRef<ViewProgressComponent>>;

  const mockEntries: Entry[] = [
    { id: 1, weight: 70, date: '01-01-2024', description: 'Start' },
    { id: 2, weight: 68, date: '15-01-2024', description: 'Progress' },
    { id: 3, weight: 65, date: '01-02-2024', description: 'Goal reached' },
  ];

  const initialState = {
    data: [],
  };

  beforeEach(async () => {
    mockDialogRef = {
      close: jest.fn(),
      afterClosed: jest.fn().mockReturnValue(of(true)),
      beforeClosed: jest.fn().mockReturnValue(of(true)),
    } as any;

    await TestBed.configureTestingModule({
      imports: [ViewProgressComponent],
      providers: [provideMockStore({ initialState })],
    }).compileComponents();

    store = TestBed.inject(MockStore);
    fixture = TestBed.createComponent(ViewProgressComponent);
    component = fixture.componentInstance;

    fixture.componentRef.setInput('dialogRef', signal(mockDialogRef));
  });

  afterEach(() => {
    fixture.destroy();
    jest.clearAllMocks();
  });

  describe('Component Creation', () => {
    it('should create', () => {
      expect(component).toBeTruthy();
    });

    it('should match snapshot', () => {
      fixture.detectChanges();
      expect(fixture).toMatchSnapshot();
    });
  });

  describe('Initialization', () => {
    it('should initialize chart options on construction', () => {
      expect(component.weightChartOptions).toBeDefined();
      expect(component.weightChartOptions.responsive).toBe(true);
      expect(component.weightChartOptions.plugins.legend.display).toBe(true);
      expect(component.weightChartOptions.plugins.tooltip.enabled).toBe(true);
    });

    it('should have null chart data initially', () => {
      expect(component.weightChartData).toBeNull();
    });

    it('should have CHART_CONFIG constants defined', () => {
      expect((component as any).CHART_CONFIG).toBeDefined();
      expect((component as any).CHART_CONFIG.BORDER_COLOR).toBe('#42A5F5');
      expect((component as any).CHART_CONFIG.TENSION).toBe(0.3);
      expect((component as any).CHART_CONFIG.POINT_RADIUS).toBe(5);
      expect((component as any).CHART_CONFIG.WEIGHT_LABEL).toBe('Weight (kg)');
      expect((component as any).CHART_CONFIG.DATE_LABEL).toBe('Date');
    });
  });

  describe('Data Loading', () => {
    beforeEach(() => {
      store.overrideSelector(reducers.getDataState, mockEntries);
      fixture.detectChanges();
    });

    it('should load and sort data from store', () => {
      expect(component.weightChartData).not.toBeNull();
      expect(component.weightChartData!.labels).toEqual([
        '01-01-2024',
        '15-01-2024',
        '01-02-2024',
      ]);
      expect(component.weightChartData!.datasets[0].data).toEqual([70, 68, 65]);
    });

    it('should create proper chart data structure', () => {
      expect(component.weightChartData!.datasets).toHaveLength(1);
      expect(component.weightChartData!.datasets[0].label).toBe('Weight (kg)');
      expect(component.weightChartData!.datasets[0].fill).toBe(false);
      expect(component.weightChartData!.datasets[0].borderColor).toBe(
        '#42A5F5'
      );
      expect(component.weightChartData!.datasets[0].tension).toBe(0.3);
      expect(component.weightChartData!.datasets[0].pointRadius).toBe(5);
    });
  });

  describe('Date Sorting', () => {
    it('should sort entries by date chronologically', () => {
      const unsortedEntries: Entry[] = [
        { id: 3, weight: 65, date: '01-02-2024', description: 'Latest' },
        { id: 1, weight: 70, date: '01-01-2024', description: 'Earliest' },
        { id: 2, weight: 68, date: '15-01-2024', description: 'Middle' },
      ];

      store.overrideSelector(reducers.getDataState, unsortedEntries);
      fixture.detectChanges();

      expect(component.weightChartData!.labels).toEqual([
        '01-01-2024',
        '15-01-2024',
        '01-02-2024',
      ]);
    });

    it('should handle different date formats correctly', () => {
      const mixedDateEntries: Entry[] = [
        { id: 1, weight: 70, date: '31-12-2023', description: 'New Year' },
        { id: 2, weight: 68, date: '01-01-2024', description: 'New Year+1' },
      ];

      store.overrideSelector(reducers.getDataState, mixedDateEntries);
      fixture.detectChanges();

      expect(component.weightChartData!.labels).toEqual([
        '31-12-2023',
        '01-01-2024',
      ]);
    });
  });

  describe('Private Methods', () => {
    it('should parseDate correctly', () => {
      const parseDate = (component as any).parseDate.bind(component);
      const date = parseDate('15-03-2024');

      expect(date.getFullYear()).toBe(2024);
      expect(date.getMonth()).toBe(2);
      expect(date.getDate()).toBe(15);
    });

    it('should sortEntriesByDate correctly', () => {
      const unsortedEntries: Entry[] = [
        { id: 2, weight: 68, date: '15-01-2024', description: 'Middle' },
        { id: 1, weight: 70, date: '01-01-2024', description: 'Earliest' },
        { id: 3, weight: 65, date: '01-02-2024', description: 'Latest' },
      ];

      const sortEntriesByDate = (component as any).sortEntriesByDate.bind(
        component
      );
      const sorted = sortEntriesByDate(unsortedEntries);

      expect(sorted.map((entry: Entry) => entry.date)).toEqual([
        '01-01-2024',
        '15-01-2024',
        '01-02-2024',
      ]);
    });

    it('should createChartData with correct structure', () => {
      const createChartData = (component as any).createChartData.bind(
        component
      );
      const chartData = createChartData(mockEntries);

      expect(chartData.labels).toEqual([
        '01-01-2024',
        '15-01-2024',
        '01-02-2024',
      ]);
      expect(chartData.datasets).toHaveLength(1);
      expect(chartData.datasets[0].data).toEqual([70, 68, 65]);
    });
  });

  describe('Empty Data Handling', () => {
    it('should handle empty data array', () => {
      store.overrideSelector(reducers.getDataState, []);
      fixture.detectChanges();

      expect(component.weightChartData).toBeNull();
    });
  });

  describe('Chart Configuration', () => {
    it('should have correct chart options structure', () => {
      const options = component.weightChartOptions;

      expect(options.scales.x.title.display).toBe(true);
      expect(options.scales.x.title.text).toBe('Date');
      expect(options.scales.y.title.display).toBe(true);
      expect(options.scales.y.title.text).toBe('Weight (kg)');
    });

    it('should match chart options snapshot', () => {
      expect(component.weightChartOptions).toMatchSnapshot();
    });
  });

  describe('Component Cleanup', () => {
    it('should use DestroyRef for cleanup', () => {
      const destroyRef = (component as any).destroyRef;
      expect(destroyRef).toBeDefined();
    });
  });
});
