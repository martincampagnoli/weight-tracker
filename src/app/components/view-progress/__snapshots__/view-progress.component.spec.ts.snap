// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewProgressComponent Chart Configuration should match chart options snapshot 1`] = `
{
  "plugins": {
    "legend": {
      "display": true,
    },
    "tooltip": {
      "enabled": true,
    },
  },
  "responsive": true,
  "scales": {
    "x": {
      "title": {
        "display": true,
        "text": "Date",
      },
    },
    "y": {
      "title": {
        "display": true,
        "text": "Weight (kg)",
      },
    },
  },
}
`;

exports[`ViewProgressComponent Component Creation should match snapshot 1`] = `
<app-view-progress
  CHART_CONFIG={[Function Object]}
  destroyRef={[Function NodeInjectorDestroyRef]}
  store={[Function _MockStore]}
  weightChartData="null"
  weightChartOptions={[Function Object]}
>
  <p-chart
    class="p-chart"
    pc2=""
    style="display: block; position: relative;"
    styleclass="w-full"
    type="line"
  >
    <canvas
      role="img"
    />
  </p-chart>
</app-view-progress>
`;
